src/dbManager.js

```diff
@@ -0,0 +1,136 @@
+
+module.exports = class DBManager {
+  constructor (db, common) {
+    this._db = db
+    this._common = common
+    this._cache = {
+      td: new Cache({ max: 1024 }),
+      header: new Cache({ max: 512 }),
+      body: new Cache({ max: 256 }),
+      numberToHash: new Cache({ max: 2048 }),
+      hashToNumber: new Cache({ max: 2048 })
+    }
+  }
+
+  getHeads () {
+    return this.get(headsKey, { valueEncoding: 'json' })
+  }
+
+  getHeadHeader () {
+    return this.get(headHeaderKey)
+  }
+
+  getHeadBlock () {
+    return this.get(headBlockKey)
+  }
+
+  async get (key, opts = {}) {
+    const dbOpts = {
+      keyEncoding: opts.keyEncoding || 'binary',
+      valueEncoding: opts.valueEncoding || 'binary'
+    }
+
+    if (opts.cache) {
+      if (!this._cache[opts.cache]) {
+        throw new Error(`Invalid cache: ${opts.cache}`)
+      }
+
+      let value = this._cache[opts.cache].get(key)
+      if (!value) {
+        value = await this._db.get(key, dbOpts)
+        this._cache[opts.cache].set(key, value)
+      }
+
+      return value
+    }
+
+    return this._db.get(key, dbOpts)
+  }
+
+  batch (ops) {
+    return this._db.batch(ops)
+  }
+
+  /**
+   * Performs a block hash to block number lookup
+   * @method _hashToNumber
+   */
+  async _hashToNumber (hash) {
+    const key = hashToNumberKey(hash)
+    return new BN(await this.get(key, { cache: 'hashToNumber' }))
+  }
+
+  /**
+   * Performs a block number to block hash lookup
+   * @method _numberToHash
+   */
+  _numberToHash (number) {
+    if (number.ltn(0)) {
+      throw new level.errors.NotFoundError()
+    }
+
+    const key = numberToHashKey(number)
+    return this.get(key, { cache: 'numberToHash' })
+  }
+
+  async _getTd (hash, number) {
+    const key = tdKey(number, hash)
+    const td = await this.get(key, { cache: 'td' })
+    return new BN(rlp.decode(td))
+  }
+
+  async _getBody (hash, number) {
+    const key = bodyKey(number, hash)
+    return rlp.decode(await this.get(key, { cache: 'body' }))
+  }
+
+  async _getHeader (hash, number) {
+    const key = headerKey(number, hash)
+    let encodedHeader = await this.get(key, { cache: 'header' })
+    return new Block.Header(rlp.decode(encodedHeader), { common: this._common })
+  }
+
+  async _getBlock (blockTag) {
+    // determine BlockTag type
+    if (Number.isInteger(blockTag)) {
+      blockTag = new BN(blockTag)
+    }
+
+    let number
+    let hash
+    if (Buffer.isBuffer(blockTag)) {
+      hash = blockTag
+      number = await this._hashToNumber(blockTag)
+    } else if (BN.isBN(blockTag)) {
+      number = blockTag
+      hash = await this._numberToHash(blockTag)
+    } else {
+      throw new Error('Unknown blockTag type')
+    }
+
+    const header = (await this._getHeader(hash, number)).raw
+    let body
+    try {
+      body = await this._getBody(hash, number)
+    } catch (e) {
+      body = [[], []]
+    }
+
+    return new Block([header].concat(body), {common: this._common})
+  }
+}
```

Comment:
Can you add JSDoc comments to the publicly exposed functions?

Reply:
(hmm, second thought: or maybe we'll leave it until integrating auto-generated documentation, up to you)

Reply:
Just tried to add it but `standard` complained: `Redundant use of `await` on a return value`

Reply:
Will try to add a bit at least for now. This made me think, should all these methods be public or private? On `Blockchain` it makes sense to have them as private, but they're the core functionality of `DBManager` :thinking: What do you think?

Reply:
Ah, I am also getting unsure on various fronts. Wasn't getting aware that ``DBManager`` isn't really prepared to go into the public space (right?). Are all publicly documented API methods from [the README](https://github.com/ethereumjs/ethereumjs-blockchain#api) still kept in the main `Blockchain` class?

Reply:
It doesn't go public by default, when users do `const Blockchain = require('ethereumjs-blockchain')` they'd get the `Blockchain` class, and they'd have to explicitly import `DBManager` to get access to these methods. Should be fine.
Yeah, the API hasn't changed and the methods in `Blockchain` only act as a wrapper for methods in `DBManager`.